lib/sqewer/worker.rb

```diff
@@ -94,22 +97,28 @@ def start
 
     # Create the provider thread. When the execution queue is exhausted,
     # grab new messages and place them on the local queue.
+    this = self # self won't be self anymore in the thread
     provider = Thread.new do
       loop do
-        break if stopping?
-
-        if queue_has_capacity?
-          messages = @connection.receive_messages
-          if messages.any?
-            messages.each {|m| @execution_queue << m }
-            @logger.debug { "[worker] Received and buffered %d messages" % messages.length } if messages.any?
+        begin
+          break if stopping?
+
+          if queue_has_capacity?
+            messages = @connection.receive_messages
+            if messages.any?
+              messages.each {|m| @execution_queue << m }
+              @logger.debug { "[worker] Received and buffered %d messages" % messages.length } if messages.any?
+            else
+              @logger.debug { "[worker] No messages received" }
+              Thread.pass
+            end
           else
-            @logger.debug { "[worker] No messages received" }
-            Thread.pass
+            @logger.debug { "[worker] Cache is full (%d items), postponing receive" % @execution_queue.length }
+            sleep SLEEP_SECONDS_ON_EMPTY_QUEUE
           end
-        else
-          @logger.debug { "[worker] Cache is full (%d items), postponing receive" % @execution_queue.length }
-          sleep SLEEP_SECONDS_ON_EMPTY_QUEUE
+        rescue StandardError => e
+          @logger.fatal "Exiting because message receiving thread died. Exception causing this: #{e.inspect}"
+          this.stop # allow any queues and/or running jobs to complete
```

Comment:
I don't know if `stop` is safely reentrant but it's a bit of an edge case. Maybe `this` can be better named `owning_worker` or something because by the time you get to read `this.stop` you already forgot what `this` is  (long methods right?)